### DISCUSSION TITLE
Is it possible to add id for each xml_node?

### DISCUSSION THREAD

**Discussion by little-brother** [zeux/pugixml#415] [General]
Currently each node has unique `hash_value`. But this value changes between runs.
Is it possible to add `id` that will be unique for each node and persistent for every runs (same input XMLs produce same ids)?
Or at least to add a user-data field?

I work on SQLite [XML extension](https://github.com/little-brother/sqlite-gui/blob/master/extensions/xml.cpp). ids will allow to connect results between different sql-XML-statements for the same XML and recreate a XML-tree in a database.

**Comment by PhantomzBack** [comment]
Yes, I think that's possible. You can use XPath and get the corresponding result `xpath_node_set` for `//*`. Then use the iterator, and create a variable to keep the count which you increment everytime. And that's the value, so you can do use the `set_value` function.

**Reply by little-brother** [reply]
`set_value` changes a node, isn't? I don't want to change XML-tree. I need a special place in the xml-node to store the counter e.g. `id` or `userdata` field. Also your suggestion requires additional parsing. I think it should be way to set the `counter`-value in a `xml_node`-constructor.

**Comment by zeux** [comment]
I'd recommend to use something like `std::unordered_map<pugi::xml_node, int>` to store external data, and fill it using a depth-first traversal after loading the document (which guarantees stability). Or use a better (faster) unordered container like Google's DenseHash or equivalent.

The reason why pugixml doesn't provide a way to store user data in the node is simple: it's a significant memory overhead which 95% of use cases don't need.

**Reply by little-brother** [reply]
Thanks.
